File: src/shared_toolkit/ui/widgets/composite/toast.py
from PyQt6.QtCore import Qt, QTimer
from PyQt6.QtGui import QFontMetrics
from PyQt6.QtWidgets import (
    QFrame,
    QHBoxLayout,
    QLabel,
    QPushButton,
    QProgressBar,
    QSizePolicy,
    QVBoxLayout,
    QWidget,
)

from shared_toolkit.ui.managers.theme_manager import ThemeManager

class ToastNotification(QWidget):
    _MARGINS_WITH_ACTION = (12, 10, 12, 10)
    _MARGINS_NO_ACTION = (12, 10, 12, 6)

    def __init__(self, parent=None):
        super().__init__(parent)

        self.setObjectName("ToastNotification")

        self.setAttribute(Qt.WidgetAttribute.WA_StyledBackground, True)

        self._on_action = None
        self._hide_timer = QTimer(self)
        self._hide_timer.setSingleShot(True)
        self._hide_timer.timeout.connect(self.hide_and_close)

        self.root_layout = QVBoxLayout(self)
        self.root_layout.setContentsMargins(0, 0, 0, 0)
        self.root_layout.setSpacing(0)

        self.content_widget = QWidget(self)
        self.main_layout = QVBoxLayout(self.content_widget)
        self.main_layout.setContentsMargins(12, 10, 12, 10)
        self.main_layout.setSpacing(8)

        self.message_label = QLabel()
        self.message_label.setWordWrap(True)
        self.message_label.setSizePolicy(
            QSizePolicy.Policy.Expanding,
            QSizePolicy.Policy.Preferred,
        )
        self.main_layout.addWidget(self.message_label)

        self.action_row = QFrame(self.content_widget)
        self.action_row_layout = QHBoxLayout(self.action_row)
        self.action_row_layout.setContentsMargins(0, 0, 0, 0)
        self.action_row_layout.setSpacing(0)

        self.action_button = QPushButton()
        self.action_button.hide()
        self.action_button.clicked.connect(self._handle_action_clicked)
        self.action_row_layout.addWidget(self.action_button, 0, Qt.AlignmentFlag.AlignLeft)
        self.action_row_layout.addStretch(1)
        self.action_row.hide()
        self.main_layout.addWidget(self.action_row)
        self.root_layout.addWidget(self.content_widget)

        self.progress_container = QWidget(self)
        self.progress_layout = QVBoxLayout(self.progress_container)
        self.progress_layout.setContentsMargins(12, 0, 12, 10)
        self.progress_layout.setSpacing(0)

        self.progress_bar = QProgressBar(self.progress_container)
        self.progress_bar.setObjectName("ToastProgressBar")
        self.progress_bar.setTextVisible(False)
        self.progress_bar.setFixedHeight(6)
        self.progress_bar.setRange(0, 100)
        self.progress_layout.addWidget(self.progress_bar)
        self.progress_container.hide()
        self.root_layout.addWidget(self.progress_container)

        self.theme_manager = ThemeManager.get_instance()

        self.theme_manager.theme_changed.connect(self._on_theme_changed)

    def _apply_content_layout_state(self):
        if self.action_row.isVisible():
            left, top, right, bottom = self._MARGINS_WITH_ACTION
            self.main_layout.setSpacing(8)
        else:
            left, top, right, bottom = self._MARGINS_NO_ACTION
            self.main_layout.setSpacing(0)
        self.main_layout.setContentsMargins(left, top, right, bottom)

    def _on_theme_changed(self):

        self.style().unpolish(self)
        self.style().polish(self)
        self.adjustSize()

    def show_message(
        self,
        message: str,
        max_width: int,
        duration: int = 3000,
        action_text: str | None = None,
        on_action=None,
        progress: int | None = None,
    ):

        self._on_action = on_action

        if action_text:
            self.action_button.setText(action_text)
            self.action_button.show()
            self.action_row.show()
        else:
            self.action_button.hide()
            self.action_row.hide()
        self._apply_content_layout_state()

        self._set_message_text(message, max_width)
        self._set_progress(progress)
        self.adjustSize()
        self.show()
        self._apply_duration(duration)

    def update_message(
        self,
        new_message: str,
        max_width: int,
        success: bool,
        duration: int = 4000,
        progress: int | None = None,
    ):
        self._apply_content_layout_state()
        self._set_message_text(new_message, max_width)
        self._set_progress(progress)
        self.adjustSize()
        self._apply_duration(duration)

    def _set_message_text(self, message: str, max_width: int):
        safe_max_width = max(180, int(max_width))
        content_margins = self.main_layout.contentsMargins()
        progress_margins = self.progress_layout.contentsMargins()
        button_width = self.action_button.sizeHint().width() if self.action_row.isVisible() else 0
        text_width = max(
            80,
            safe_max_width
            - content_margins.left()
            - content_margins.right()
        )
        font_metrics = QFontMetrics(self.message_label.font())
        lines = message.split("\n") if "\n" in message else [message]
        longest_line_width = max((font_metrics.horizontalAdvance(line) for line in lines), default=0)
        desired_text_width = max(80, min(text_width, longest_line_width + 4))
        desired_toast_width = min(
            safe_max_width,
            max(
                180,
                desired_text_width
                + content_margins.left()
                + content_margins.right()
                + 4,
                button_width
                + content_margins.left()
                + content_margins.right(),
                desired_text_width
                + progress_margins.left()
                + progress_margins.right(),
            ),
        )
        final_text_width = max(
            80,
            desired_toast_width
            - content_margins.left()
            - content_margins.right()
        )

        self.setFixedWidth(desired_toast_width)
        self.content_widget.setFixedWidth(desired_toast_width)
        self.progress_container.setFixedWidth(desired_toast_width)
        self.message_label.setFixedWidth(final_text_width)
        self.message_label.setText(message)
        self.action_row.setFixedWidth(final_text_width)
        self.message_label.updateGeometry()
        self.content_widget.adjustSize()
        self.adjustSize()
        self.updateGeometry()

    def _set_progress(self, progress: int | None):
        if progress is None:
            self.progress_container.hide()
            return

        safe_progress = max(0, min(100, int(progress)))
        self.progress_bar.setValue(safe_progress)
        self.progress_container.show()

    def _apply_duration(self, duration: int):
        self._hide_timer.stop()
        if duration > 0:
            self._hide_timer.start(duration)

    def hide_and_close(self):
        self._hide_timer.stop()
        self.hide()
        self.close()

    def _handle_action_clicked(self):
        try:
            if callable(self._on_action):
                self._on_action()
        finally:
            self.hide_and_close()

    def mousePressEvent(self, event):

        if self.action_button.isVisible() and self.action_button.geometry().contains(
            event.pos()
        ):
            return super().mousePressEvent(event)
        self.hide_and_close()
        event.accept()

from PyQt6.QtCore import QEvent, QObject, QPoint, QRect

class ToastManager(QObject):
    def __init__(self, parent_window: QWidget, image_label_widget: QWidget):
        super().__init__(parent_window)
        self.parent_window = parent_window
        self.image_label = image_label_widget
        self.toasts = {}
        self.next_toast_id = 0
        self.spacing = 10

        self.parent_window.installEventFilter(self)
        if self.image_label is not None:
            self.image_label.installEventFilter(self)

    def show_toast(
        self,
        message: str,
        duration: int = 0,
        action_text: str | None = None,
        on_action=None,
        progress: int | None = None,
    ) -> int:
        toast_id = self.next_toast_id
        self.next_toast_id += 1

        toast = ToastNotification(self.parent_window)
        self.toasts[toast_id] = toast

        max_toast_width = (
            int(self.image_label.width() * 0.4) if self.image_label else 400
        )
        toast.show_message(
            message,
            max_toast_width,
            duration,
            action_text=action_text,
            on_action=on_action,
            progress=progress,
        )

        toast.destroyed.connect(lambda tid=toast_id: self.toasts.pop(tid, None))

        QTimer.singleShot(0, self._reposition_toasts)
        return toast_id

    def update_toast(
        self,
        toast_id: int,
        new_message: str,
        success: bool,
        duration: int = 4000,
        progress: int | None = None,
    ):
        if toast_id in self.toasts:
            max_toast_width = (
                int(self.image_label.width() * 0.4) if self.image_label else 400
            )
            toast = self.toasts[toast_id]

            toast._on_action = None
            toast.action_button.hide()
            toast.update_message(
                new_message,
                max_toast_width,
                success,
                duration,
                progress=progress,
            )
            self._reposition_toasts()

    def close_toast(self, toast_id: int):
        toast = self.toasts.get(toast_id)
        if toast is None:
            return
        toast.hide_and_close()

    def _reposition_toasts(self):
        if not self.parent_window:
            return

        anchor_point = QPoint(0, 0)
        if self.image_label is not None:
            anchor_point = self.image_label.mapTo(self.parent_window, QPoint(0, 0))

        at_x = anchor_point.x() + self.spacing
        at_y = anchor_point.y() + self.spacing

        for toast in list(self.toasts.values()):
            if not toast.isVisible():
                continue

            toast.setGeometry(QRect(at_x, at_y, toast.width(), toast.height()))
            toast.raise_()

            at_y += toast.height() + self.spacing

    def eventFilter(self, watched, event):
        if watched in (self.parent_window, self.image_label) and event.type() in [
            QEvent.Type.Resize,
            QEvent.Type.Move,
            QEvent.Type.Show,
            QEvent.Type.WindowStateChange,
            QEvent.Type.LayoutRequest,
        ]:
            QTimer.singleShot(0, self._reposition_toasts)
        return False

File: cpp/toolkit/src/composite/toast.cpp
#include "sli/toolkit/composite/toast.h"

#include <QEvent>
#include <QFontMetrics>
#include <QHBoxLayout>
#include <QLabel>
#include <QMouseEvent>
#include <QPainter>
#include <QProgressBar>
#include <QStyle>
#include <QVBoxLayout>

#include <algorithm>

#include "sli/toolkit/buttons/button.h"
#include "sli/toolkit/theme.h"

namespace sli::toolkit {

// ===========================================================================
// ToastWidget
// ===========================================================================

ToastWidget::ToastWidget(QWidget* parent)
    : QWidget(parent) {
  Q_ASSERT(parent != nullptr);

  setObjectName(QStringLiteral("ToastNotification"));
  setWindowFlags(Qt::Widget);
  setAttribute(Qt::WA_StyledBackground, false);

  hideTimer_ = new QTimer(this);
  hideTimer_->setSingleShot(true);
  connect(hideTimer_, &QTimer::timeout, this, &ToastWidget::hideAndClose);

  rootLayout_ = new QVBoxLayout(this);
  rootLayout_->setContentsMargins(0, 0, 0, 0);
  rootLayout_->setSpacing(0);

  contentWidget_ = new QWidget(this);
  contentWidget_->setObjectName(QStringLiteral("ToastContentWidget"));
  contentWidget_->setAttribute(Qt::WA_StyledBackground, false);
  contentWidget_->setAttribute(Qt::WA_TranslucentBackground, true);

  mainLayout_ = new QVBoxLayout(contentWidget_);
  mainLayout_->setContentsMargins(12, 10, 12, 10);
  mainLayout_->setSpacing(8);

  messageLabel_ = new QLabel();
  messageLabel_->setWordWrap(true);
  messageLabel_->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Preferred);
  mainLayout_->addWidget(messageLabel_);

  actionRow_ = new QWidget(contentWidget_);
  actionRowLayout_ = new QHBoxLayout(actionRow_);
  actionRowLayout_->setContentsMargins(0, 0, 0, 0);
  actionRowLayout_->setSpacing(6);
  actionRowLayout_->addStretch(1);
  actionRow_->hide();
  mainLayout_->addWidget(actionRow_);

  rootLayout_->addWidget(contentWidget_);

  progressContainer_ = new QWidget(this);
  progressContainer_->setObjectName(QStringLiteral("ToastProgressContainer"));
  progressContainer_->setAttribute(Qt::WA_StyledBackground, false);
  progressContainer_->setAttribute(Qt::WA_TranslucentBackground, true);

  progressLayout_ = new QVBoxLayout(progressContainer_);
  progressLayout_->setContentsMargins(12, 0, 12, 10);
  progressLayout_->setSpacing(0);

  progressBar_ = new QProgressBar(progressContainer_);
  progressBar_->setObjectName(QStringLiteral("ToastProgressBar"));
  progressBar_->setTextVisible(false);
  progressBar_->setFixedHeight(6);
  progressBar_->setRange(0, 100);
  progressLayout_->addWidget(progressBar_);
  progressContainer_->hide();
  rootLayout_->addWidget(progressContainer_);

  Theme::onThemeChanged(this, [this]() { onThemeChanged(); });
  applySurfaceState();
}

void ToastWidget::onThemeChanged() {
  repolishSurfaceWidgets();
  adjustSize();
}

void ToastWidget::repolishSurfaceWidgets() {
  QList<QWidget*> widgets = {this, contentWidget_, progressContainer_,
                             progressBar_};
  for (QWidget* w : widgets) {
    if (w == nullptr) continue;
    QStyle* s = w->style();
    s->unpolish(w);
    s->polish(w);
    w->update();
  }
}

void ToastWidget::applySurfaceState() {
  bool hasProgress = !progressContainer_->isHidden();
  contentWidget_->setProperty("hasProgress", hasProgress);
  progressContainer_->setProperty("hasProgress", hasProgress);
  repolishSurfaceWidgets();
}

void ToastWidget::applyContentLayoutState() {
  if (actionRow_->isVisible()) {
    mainLayout_->setContentsMargins(kMarginsWithAction[0],
                                    kMarginsWithAction[1],
                                    kMarginsWithAction[2],
                                    kMarginsWithAction[3]);
    mainLayout_->setSpacing(8);
  } else {
    mainLayout_->setContentsMargins(kMarginsNoAction[0],
                                    kMarginsNoAction[1],
                                    kMarginsNoAction[2],
                                    kMarginsNoAction[3]);
    mainLayout_->setSpacing(0);
  }
}

void ToastWidget::showMessage(const QString& content, int maxWidth,
                              int duration,
                              const QList<ToastConfig>& actions,
                              std::optional<int> progress) {
  setActions(actions);
  applyContentLayoutState();
  setContent(content, maxWidth);
  setProgress(progress);
  adjustSize();
  show();
  applyDuration(duration);
}

void ToastWidget::updateMessage(const QString& content, int maxWidth,
                                bool success, int duration,
                                const QList<ToastConfig>& actions,
                                std::optional<int> progress) {
  if (!actions.isEmpty()) {
    setActions(actions);
  }
  applyContentLayoutState();
  if (!content.isEmpty()) {
    setContent(content, maxWidth);
  } else {
    fitToContent(maxWidth);
  }
  setProgress(progress);
  adjustSize();
  applyDuration(duration);
}

void ToastWidget::setActions(const QList<ToastConfig>& actions) {
  // Remove existing action widgets (keep the stretch at index count()-1)
  while (actionRowLayout_->count() > 1) {
    QLayoutItem* item = actionRowLayout_->takeAt(0);
    if (QWidget* w = item->widget()) {
      w->setParent(nullptr);
      w->deleteLater();
    }
    delete item;
  }
  actionWidgets_.clear();

  for (const ToastConfig& action : actions) {
    if (action.text.isEmpty()) continue;

    Button::Config cfg;
    cfg.text = action.text;
    cfg.icon = action.icon;
    cfg.variant = Button::Variant::Surface;
    cfg.size = QSize(0, 28);
    cfg.density = QStringLiteral("compact");

    auto* button = new Button(cfg, actionRow_);
    connect(button, &Button::clicked, this,
            [this, cb = action.callback, dismiss = action.dismiss]() {
              handleActionClicked(cb, dismiss);
            });

    actionRowLayout_->insertWidget(actionRowLayout_->count() - 1, button, 0,
                                   Qt::AlignLeft);
    actionWidgets_.append(button);
  }

  actionRow_->setVisible(!actionWidgets_.isEmpty());
}

void ToastWidget::setContent(const QString& content, int maxWidth) {
  // The Python version handles QWidget content separately; here content is
  // always a string since C++ doesn't have Python's dynamic typing on this
  // parameter. The string path is the primary use case.
  setMessageText(content, maxWidth);
}

void ToastWidget::setMessageText(const QString& message, int maxWidth) {
  if (customContent_ != nullptr) {
    mainLayout_->removeWidget(customContent_);
    customContent_->setParent(nullptr);
    customContent_->deleteLater();
    customContent_ = nullptr;
  }
  messageLabel_->show();

  int safeMaxWidth = qMax(180, maxWidth);
  QMargins contentMargins = mainLayout_->contentsMargins();
  QMargins progressMargins = progressLayout_->contentsMargins();
  int actionsWidth =
      actionRow_->isVisible() ? actionRow_->sizeHint().width() : 0;

  int textWidth = qMax(80, safeMaxWidth - contentMargins.left() -
                                contentMargins.right());

  QFontMetrics fm(messageLabel_->font());
  QStringList lines = message.contains(QLatin1Char('\n'))
                          ? message.split(QLatin1Char('\n'))
                          : QStringList{message};
  int longestLineWidth = 0;
  for (const QString& line : lines) {
    longestLineWidth = qMax(longestLineWidth, fm.horizontalAdvance(line));
  }
  int desiredTextWidth =
      qMax(80, qMin(textWidth, longestLineWidth + 4));
  int desiredToastWidth = qMin(
      safeMaxWidth,
      std::max({180,
                desiredTextWidth + contentMargins.left() +
                    contentMargins.right() + 4,
                actionsWidth + contentMargins.left() + contentMargins.right(),
                desiredTextWidth + progressMargins.left() +
                    progressMargins.right()}));
  int finalTextWidth = qMax(
      80, desiredToastWidth - contentMargins.left() - contentMargins.right());

  setFixedWidth(desiredToastWidth);
  contentWidget_->setFixedWidth(desiredToastWidth);
  progressContainer_->setFixedWidth(desiredToastWidth);
  messageLabel_->setFixedWidth(finalTextWidth);
  messageLabel_->setText(message);
  actionRow_->setFixedWidth(finalTextWidth);
  messageLabel_->updateGeometry();
  contentWidget_->adjustSize();
  adjustSize();
  updateGeometry();
}

void ToastWidget::fitToContent(int maxWidth) {
  int safeMaxWidth = qMax(180, maxWidth);
  QMargins contentMargins = mainLayout_->contentsMargins();
  QMargins progressMargins = progressLayout_->contentsMargins();

  int contentWidth =
      customContent_ != nullptr
          ? customContent_->sizeHint().width()
          : messageLabel_->sizeHint().width();
  int actionsWidth =
      actionRow_->isVisible() ? actionRow_->sizeHint().width() : 0;

  int desiredToastWidth = qMin(
      safeMaxWidth,
      std::max({180,
                contentWidth + contentMargins.left() + contentMargins.right(),
                actionsWidth + contentMargins.left() + contentMargins.right(),
                contentWidth + progressMargins.left() +
                    progressMargins.right()}));

  setFixedWidth(desiredToastWidth);
  contentWidget_->setFixedWidth(desiredToastWidth);
  progressContainer_->setFixedWidth(desiredToastWidth);
  actionRow_->setFixedWidth(
      qMax(80, desiredToastWidth - contentMargins.left() -
                   contentMargins.right()));
  contentWidget_->adjustSize();
  adjustSize();
  updateGeometry();
}

void ToastWidget::setProgress(std::optional<int> progress) {
  if (!progress.has_value()) {
    progressContainer_->hide();
    applySurfaceState();
    return;
  }
  int safeProgress = qBound(0, progress.value(), 100);
  progressBar_->setValue(safeProgress);
  progressContainer_->show();
  applySurfaceState();
}

void ToastWidget::applyDuration(int duration) {
  hideTimer_->stop();
  if (duration > 0) {
    hideTimer_->start(duration);
  }
}

void ToastWidget::hideAndClose() {
  hideTimer_->stop();
  hide();
  close();
}

void ToastWidget::handleActionClicked(std::function<void()> callback,
                                      bool dismiss) {
  if (callback) {
    callback();
  }
  if (dismiss) {
    hideAndClose();
  }
}

void ToastWidget::paintEvent(QPaintEvent* event) {
  QPainter painter(this);
  painter.setRenderHint(QPainter::Antialiasing);
  QRect rect = this->rect().adjusted(0, 0, -1, -1);
  painter.setBrush(QBrush(Theme::getColor(QStringLiteral("toast.background"))));
  painter.setPen(QPen(Theme::getColor(QStringLiteral("toast.border")), 1));
  painter.drawRoundedRect(rect, 8, 8);
  painter.end();
  QWidget::paintEvent(event);
}

void ToastWidget::mousePressEvent(QMouseEvent* event) {
  if (actionRow_->isVisible() &&
      actionRow_->rect().contains(actionRow_->mapFrom(this, event->pos()))) {
    QWidget::mousePressEvent(event);
    return;
  }
  hideAndClose();
  event->accept();
}

// ===========================================================================
// ToastManager
// ===========================================================================

ToastManager::ToastManager(QWidget* parentWindow, QWidget* imageLabel)
    : QObject(parentWindow) {
  QWidget* hostParent = parentWindow;
  if (hostParent == nullptr && imageLabel != nullptr) {
    hostParent = imageLabel->window();
  }
  Q_ASSERT(hostParent != nullptr);

  parentWindow_ = hostParent;
  imageLabel_ = imageLabel;

  if (parentWindow_ != nullptr) {
    parentWindow_->installEventFilter(this);
  }
  if (imageLabel_ != nullptr) {
    imageLabel_->installEventFilter(this);
  }
}

int ToastManager::showToast(const QString& content, int duration,
                            const QList<ToastConfig>& actions,
                            std::optional<int> progress, bool success) {
  int toastId = nextId_++;
  auto* toast = new ToastWidget(parentWindow_);
  toast->setProperty("toastSuccess", success);
  toasts_[toastId] = toast;

  connect(toast, &QObject::destroyed, this,
          [this, toastId]() { toasts_.remove(toastId); });

  toast->showMessage(content, toastMaxWidth(), duration, actions, progress);
  positionToasts();
  toast->show();
  toast->raise();
  QTimer::singleShot(0, this, &ToastManager::positionToasts);
  return toastId;
}

void ToastManager::updateToast(int toastId, const QString& content,
                               bool success, int duration,
                               const QList<ToastConfig>& actions,
                               std::optional<int> progress) {
  auto it = toasts_.find(toastId);
  if (it == toasts_.end() || it.value().isNull()) return;

  ToastWidget* toast = it.value().data();
  toast->setProperty("toastSuccess", success);
  toast->updateMessage(content, toastMaxWidth(), success, duration, actions,
                       progress);
  positionToasts();
  toast->show();
  toast->raise();
  QTimer::singleShot(0, this, &ToastManager::positionToasts);
}

void ToastManager::closeToast(int toastId) {
  auto it = toasts_.find(toastId);
  if (it == toasts_.end()) return;
  if (!it.value().isNull()) {
    it.value()->hideAndClose();
  }
  toasts_.remove(toastId);
}

int ToastManager::toastMaxWidth() const {
  if (imageLabel_ != nullptr) {
    return qMax(260, static_cast<int>(imageLabel_->width() * 0.42));
  }
  if (parentWindow_ != nullptr) {
    return qMax(260, static_cast<int>(parentWindow_->width() * 0.35));
  }
  return 360;
}

void ToastManager::positionToasts() {
  if (parentWindow_.isNull()) return;

  QPoint anchorPoint(0, 0);
  if (imageLabel_ != nullptr) {
    anchorPoint = imageLabel_->mapTo(parentWindow_, QPoint(0, 0));
  }

  int atX = anchorPoint.x() + spacing_;
  int atY = anchorPoint.y() + spacing_;

  for (auto it = toasts_.begin(); it != toasts_.end(); ++it) {
    if (it.value().isNull()) continue;
    ToastWidget* toast = it.value().data();
    if (!toast->isVisible()) continue;
    toast->setGeometry(QRect(atX, atY, toast->width(), toast->height()));
    toast->raise();
    atY += toast->height() + spacing_;
  }
}

bool ToastManager::eventFilter(QObject* watched, QEvent* event) {
  if ((watched == parentWindow_.data() ||
       watched == imageLabel_.data()) &&
      (event->type() == QEvent::Resize ||
       event->type() == QEvent::Move ||
       event->type() == QEvent::Show ||
       event->type() == QEvent::WindowStateChange ||
       event->type() == QEvent::LayoutRequest)) {
    QTimer::singleShot(0, this, &ToastManager::positionToasts);
  }
  return false;
}

}  // namespace sli::toolkit